prometheus_client/exposition.py

```diff
@@ -307,20 +324,77 @@ def sample_line(samples):
             raise
 
         for suffix, lines in sorted(om_samples.items()):
-            output.append('# HELP {} {}\n'.format(openmetrics.escape_metric_name(metric.name + suffix),
+            output.append('# HELP {} {}\n'.format(openmetrics.escape_metric_name(metric.name + suffix, escaping),
                                                   metric.documentation.replace('\\', r'\\').replace('\n', r'\n')))
-            output.append(f'# TYPE {openmetrics.escape_metric_name(metric.name + suffix)} gauge\n')
+            output.append(f'# TYPE {openmetrics.escape_metric_name(metric.name + suffix, escaping)} gauge\n')
             output.extend(lines)
     return ''.join(output).encode('utf-8')
 
 
 def choose_encoder(accept_header: str) -> Tuple[Callable[[CollectorRegistry], bytes], str]:
+    # Python client library accepts a much narrower range of content-types than
+    # Prometheus does -- UTF-8 is only supported on OpenMetrics v1.0.0.
     accept_header = accept_header or ''
+    escaping = openmetrics.UNDERSCORES
     for accepted in accept_header.split(','):
         if accepted.split(';')[0].strip() == 'application/openmetrics-text':
-            return (openmetrics.generate_latest,
-                    openmetrics.CONTENT_TYPE_LATEST)
-    return generate_latest, CONTENT_TYPE_LATEST
+            toks = accepted.split(';')
+            version = _get_version(toks)
+            escaping = _get_escaping(toks)
+            # Only return an escaping header if we have a good version and
+            # mimetype.
+            if Version(version) >= Version('1.0.0'):
+                return (openmetrics.generate_latest_fn(escaping), 
+                        openmetrics.CONTENT_TYPE_LATEST + '; escaping=' + str(escaping))
+        elif accepted.split(';')[0].strip() == 'text/plain':
+            toks = accepted.split(';')
+            version = _get_version(toks)
+            escaping = _get_escaping(toks)
+            # Only return an escaping header if we have a good version and
+            # mimetype.
+            if Version(version) >= Version('1.0.0'):
+                return (openmetrics.generate_latest_fn(escaping),
```

Comment:
This returns an openmetrics payload wrapped inside of `text/plain` which is not correct, it needs to be calling `generate_latest` in this module instead. I'd suggest using `partial(generate_latest, escaping=escaping)` from `functools`.

Reply:
You could probably use mocks to avoid generating lines and then parsing them appropriately. See https://docs.python.org/3/library/unittest.mock.html#where-to-patch for how to patch a method in a module. Just checking for the presence of `# EOF` at the end of the output could also tell the difference between plain and openmetrics if you do generate lines.